torch/csrc/autograd/engine.cpp

```diff
@@ -72,23 +73,38 @@ struct GraphTask {
   // Notified when a task finishes executing.  Check outstanding_tasks to see
   // if all tasks are done.
   std::condition_variable not_done;
-  const Engine::pre_callback_map& pre_callbacks;
-  const Engine::post_callback_map& post_callbacks;
   std::unordered_map<Function*, InputBuffer> not_ready;
   std::unordered_map<Function*, int> dependencies;
 
+  struct ExecInfo {
+    struct Capture {
+      Capture(int input_idx, int output_idx) : input_idx(input_idx), output_idx(output_idx) {}
+      int input_idx; // within Function inputs
+      int output_idx; // within the output vector of a GraphTask
+    };
+
+    bool needed;
+    std::unique_ptr<std::vector<Capture>> captures;
+  };
+  // Exec info has a bit complicated semantics. If it's empty, it means the task is
+  // run in a "default" mode, which means that all next_functions we encounter should
+  // get executed. If it's not empty, only functions that have an entry and this entry
+  // has needed == True should be executed.
+  std::unordered_map<Function*, ExecInfo> exec_info;
+  std::vector<Variable> captured_vars;
+
+  void init_to_execute(const std::shared_ptr<Function>& graph_root, const function_list& captures);
```

Comment:
That's... completely different than everything we've been doing so far.

Reply:
Not sure where else we're passing `std::shared_ptr<Function>` as arguments without taking ownership. Herb Sutter's [recommendation](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) is:

1) Pass `std::shared_ptr<Function>` when taking ownership
2) Pass `Function&` when the argument cannot be null
3) Pass `Function*` when it may be null

Our current code that take `std::shared_ptr<Function>` generally takes ownership.